File: clojure/cars-assemble/src/cars_assemble.clj
(ns cars-assemble)

(defn rates [speed]
  (cond
    (= speed 0) 0
    (<= speed 4) 1
    (<= speed 8) 0.9
    (<= speed 9) 0.8
    (= speed 10) 0.77))

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
  (double (* (* 221 speed) (rates speed))))

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int (/ (production-rate speed) 60)))

File: go/cars-assemble/cars_assemble.go
package cars

import "math"

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) * successRate / 100
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	CarsPerHour := CalculateWorkingCarsPerHour(productionRate, successRate)
	CarsPerMinute := CarsPerHour / float64(60)
	return int(math.Floor(CarsPerMinute))
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	costOfOneCar := 10000
	groupsOfTen := carsCount / 10
	if groupsOfTen == 0 {
		return uint(carsCount) * uint(costOfOneCar)
	}
	costOfTenCars := 95000
	groupCost := groupsOfTen * costOfTenCars
	remainder := carsCount - (groupsOfTen * 10)
	remainderCost := remainder * costOfOneCar

	return uint(groupCost) + uint(remainderCost)
}
